keps/sig-network/3458-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,787 @@
+  service proxy responds to the health check probe. This KEP originally intended
+  on solving this problem by having `healthCheckNodePort` become this common
+  denominator. However, we are forced to abide to the constraint: "we can't open
+  up anymore node ports than we already do". One idea aimed at solving this was
+  to introduce a `healthCheckPath` that would be keyed per-service. That
+  involves API changes, compatibility concerns and other issues which would slow
+  down the progression of this KEP. It has therefor been left out with the
+  intention of filing a dedicated KEP for that problem. Such a proposal would
+  reduce the amount of node ports open by the service proxy even further (for
+  Kube-proxy specifically: that would be 0 since it already doesn't open any
+  `NodePort` anymore).
+
+## Proposal
+
+Up until this point the KCCM has defined the following conditions for what is
+considered a "healthy" node:
+
+1. Is the node not tainted with the key: `ToBeDeletedByClusterAutoscaler`?
+2. Is the node condition `NodeReady` equal to `ConditionTrue`?
+
+The first predicate was introduced by the
+[Kube PR #105946](https://github.com/kubernetes/kubernetes/pull/105946) which
+stated:
+
+> If multiple...nodes are being deleted, the list of nodes passed to
+> cloud-provider code via UpdateLoadBalancer() will be stale, causing a few
+> failed updates before the backends stabilize.
+
+It seemed, from reading the above, that the goal was to ensure that we get a
+head start at deleting the node from the load balancer set before the underlying
+VM instance has been completely cleaned up by the auto-scaler. The
+interpretation of the PR indicated that if the predicate is not used: we'll only
+get notified the moment the Node is deleted in the API server, which might be
+after the VM instance is GC'ed, causing the update call on the load balancer to
+fail because the VM instance cannot be found and thus, have the KCCM experience
+transient errors until the cluster converges to a "steady-state".
+
+The KCCM updates load balancers by invoking calls to
+`EnsureLoadBalancer`/`UpdateLoadBalancer` (defined by the cloud-provider API)
+which accept the argument `nodes` as the set of __all the remaining cluster
+nodes__. Even though it is not expected that deleting a single cluster Node
+would trigger the behavior mentioned on that PR, deleting multiple at the same
+time, might. The following sequence of events presents one single mode of error
+for the KCCM __if that condition is removed__.
+
+1. `Node1`, `Node2` are to be scaled down and removed by the auto-scaler
+2. `Node1`, `Node2` are tainted
+3. `Node1`, `Node2` are drained, etc...
+4. Instance associated with `Node1` is deleted
+5. `Node1` is deleted
+6. KCCM re-configures LB set with all nodes (including `Node2`)
+7. Instance associated with `Node2` is deleted
+8. KCCM starts erroring because the instance associated with `Node2` cannot be found
```

Comment:
If we think that all the clouds support automatic removal from LBs when the VM is deleted, maybe we fix this by first reading the LB set and seeing if the node is in there at all.  If not, skip.  For CA it should always be absent.

Reply:
Yes, that seems doable. Or we just accept the fact that we experience some transient errors from time to time and that we eventually will converge on the desired state. This corner case shouldn't take that long to recover from in any case